File: C++/572. Subtree of Another Tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    bool isSubtree(TreeNode* s, TreeNode* t) {
        if (s == NULL && t == NULL)
            return true;
        if (s == NULL || t == NULL)
            return false;        
        if (isSameTree(s, t))
            return true;        
        return isSubtree(s->left, t) || isSubtree(s->right, t);           
    }

private:
    bool isSameTree(TreeNode* s, TreeNode* t){
        if (s == NULL && t == NULL)
            return true;
        if (s == NULL || t == NULL)
            return false;
        if (s->val != t->val)
            return false;
        else
            return isSameTree(s->left, t->left) && isSameTree(s->right, t->right);
    }
    
};

File: Python/572. Subtree of Another Tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None


class Solution:
    def isSubtree(self, s: TreeNode, t: TreeNode) -> bool:
        def isSame(s, t):
            if s == None and t == None:
                return True
            if s == None or t == None:
                return False
            if s.val == t.val and isSame(s.left, t.left) and isSame(s.right, t.right):
                return True
        if s == None and t == None:
            return True
        if s == None or t == None:
            return False
        return isSame(s, t) or self.isSubtree(s.left, t) or self.isSubtree(s.right, t)
